libs/ledger-live-common/src/deviceSDK/tasks/updateFirmware.ts

```diff
@@ -75,44 +75,44 @@ function internalUpdateFirmwareTask({
   updateContext,
 }: UpdateFirmwareTaskArgs): Observable<UpdateFirmwareTaskEvent> {
   return new Observable((subscriber) => {
-    withTransport(deviceId)(({ transportRef }) =>
+    const sub = withTransport(deviceId)(({ transportRef }) =>
       concat(
         quitApp(transportRef.current).pipe(
           switchMap(() => waitForGetVersion(transportRef, {})),
           switchMap((value) => {
             const { firmwareInfo } = value;
 
-            // TODO: we're repeating the handling of the unresponsive event here... is there a way to make this better
-            // the problem that if we want to change multiple commands that may have unresponsive events
-            // and only continue with the next command if the event is not unresponsive we might have to handle multiple times
-            // the unresponsive event :/
-            return installOsuFirmware({
-              firmwareInfo,
-              updateContext,
-              transport: transportRef.current,
-            }).pipe(
-              map((e) => {
-                if (e.type === "unresponsive") {
-                  return {
-                    type: "error" as const,
-                    error: new LockedDeviceError(),
-                  };
-                }
+            return subscriber.closed
+              ? EMPTY
```

Comment:
It's not really about emitting an event but rather about preventing API calls to be made and APDUs to be sent in case the observable gets unsubscribed. It's a way to stop the side-effects of the observable.

Reply:
It's there for when the unsubscribe happens exactly at the point we're already inside this function

Reply:
Ok i actually agree with you, it's necessary !

I tested with:
```JS
import { Observable, interval, from } from "rxjs";
import { switchMap } from "rxjs/operators";

function asyncDelay(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

const subscription = new Observable((subscriber) => {
  interval(2000)
    .pipe(
      switchMap((value) => {
        return from(
          (async () => {
            console.log(`Start switch: ${value}`);
            await asyncDelay(1000);
            console.log(`End switch: ${value}`);
            return Promise.resolve(value);
          })()
        );
      })
    )
    .subscribe(subscriber);
}).subscribe({
  next: (value) => {
    console.log(`➡️ next: ${value}`);
    setTimeout(() => {
      console.log(`🚨 unsubscribing from ${value}`);
      subscription.unsubscribe();
    }, 1001);
  },
  complete: () => console.log(`completed`)
});
```

and got the following result:
```bash
Start switch: 0 
End switch: 0 
➡️ next: 0 
Start switch: 1 
🚨 unsubscribing from 0 
End switch: 1 
```

Showing that even if an unsubscription occurred, a block inside a `switchMap` is not stopped directly but is finished until the end (just nothing is emitted after)

Reply:
It's really cool that you did the test, that's awesome! Gonna save it for future refrence hahaha